helix-rest/src/main/java/org/apache/helix/rest/metadatastore/ShardingKeyTrieNode.java

```diff
@@ -0,0 +1,118 @@
+package org.apache.helix.rest.metadatastore;
+
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import java.util.Collections;
+import java.util.Map;
+
+/**
+ * A trie node that is used to represent Zk path sharding keys. Terminal nodes contain ZkRealm
+ * addresses.
+ */
+public class ShardingKeyTrieNode {
+  /**
+   * Children trie nodes of this trie node. Each node is identified by their name.
+   */
+  private final Map<String, ShardingKeyTrieNode> _children;
+  /**
+   * Boolean value dictating if this trie node is a terminal node.
+   */
+  private final boolean _isLeaf;
+  /**
+   * This trie node's name.
+   */
+  private final String _name;
+  /**
+   * ZkRealm address corresponding to this trie node. The value is only valid when the node is a
+   * terminal node.
+   */
+  private final String _zkRealmAddress;
+
+  public ShardingKeyTrieNode(Map<String, ShardingKeyTrieNode> children, boolean isLeaf, String name,
+      String zkRealmAddress) {
+    _children = children;
+    _isLeaf = isLeaf;
+    _name = name;
+    _zkRealmAddress = zkRealmAddress;
+  }
+
+  public Map<String, ShardingKeyTrieNode> getChildren() {
+    return Collections.unmodifiableMap(_children);
```

Comment:
1. The map is final
2. There is no set method to update the map.
Why not directly put it in an unmodifiable map in the constructor? So we don't need to create a map on every get.

Reply:
Makes sense. Let me make the update.